hicexplorer/hicBuildMatrix.py

```diff
@@ -822,74 +844,363 @@ def main(args=None):
 
         for mate in [mate1, mate2]:
             # fill in coverage vector
-            vec_start = max(0, mate.pos - mate_bin.begin) / binsize
-            vec_end = min(len(coverage[mate_bin_id]), vec_start +
-                          len(mate.seq) / binsize)
-            coverage[mate_bin_id][vec_start:vec_end] += 1
+            vec_start = max(0, mate.pos - mate_bin.begin) / pBinsize
+            length_coverage = pCoverageIndex[mate_bin_id].end - pCoverageIndex[mate_bin_id].begin
+            vec_end = min(length_coverage, vec_start +
+                          len(mate.seq) / pBinsize)
+            coverage_index = pCoverageIndex[mate_bin_id].begin + vec_start
+            coverage_end = pCoverageIndex[mate_bin_id].begin + vec_end
+            for i in xrange(coverage_index, coverage_end, 1):
+                pCoverage[i] += 1
 
         row.append(mate_bins[0])
         col.append(mate_bins[1])
         data.append(1)
 
         pair_added += 1
+        if pOutputBamSet:
+            # out_bam = pysam.Samfile(pOutputName, 'wb', template=pTemplate)
+            # prepare data for bam output
+            # set the flag to point that this data is paired
+            mate1.flag |= 0x1
+            mate2.flag |= 0x1
+
+            # set one read as the first in pair and the
+            # other as second
+            mate1.flag |= 0x40
+            mate2.flag |= 0x80
+
+            # set chrom of mate
+            mate1.mrnm = mate2.rname
+            mate2.mrnm = mate1.rname
+
+            # set position of mate
+            mate1.mpos = mate2.pos
+            mate2.mpos = mate1.pos
+            out_bam.write(mate1)
+            out_bam.write(mate2)
+
+    hic_matrix = coo_matrix((data, (row, col)), shape=(pMatrixSize, pMatrixSize), dtype='uint16')
+    row = []
+    col = []
+    data = []
+    if pOutputBamSet:
+        out_bam.close()
+
+    pQueueOut.put([hic_matrix, [one_mate_unmapped, one_mate_low_quality, one_mate_not_unique, dangling_end, self_circle, self_ligation, same_fragment,
```

Comment:
@bgruening I tried that with hiding the variables in two classes, see commit 226c411. For some reason it is slower by a factor of 2. I don't want to invest more time to find the reason, I go back to the version that was used before. Maybe this can be done in a later PR.